McAfee openDXL/ATD/MAR/ESM Bots

tux78 opened the pull request:
Hi,

I have created the following Bots:
- openDXL collector listening to openDXL for messages
- McAfee ATD parser, parsing IP, FQDN and hash out of ATD messages sent via DXL
- McAfee AR expert, querying environment for IP/FQDN/Hash value occurences via DXL/MAR 
- McAfee ESM output, adding IPs found via MAR expert bot to ESM watchlists

Regards,
Seb

wagner-certat commented on intelmq/bots/collectors/opendxl/collector.py:
```diff
@@ -0,0 +1,106 @@
+import threading
+import importlib
+
+from dxlclient.callbacks import EventCallback
```
Please put this in a try-except block and do error handling in init, see for example https://github.com/certtools/intelmq/blob/maintenance/intelmq/bots/outputs/mongodb/output.py#L8

wagner-certat commented on intelmq/bots/collectors/opendxl/collector.py:
```diff
@@ -0,0 +1,106 @@
+        else:
+            self.logger.info("There was no openDXL client I could stop.")
+
+if __name__ == "__main__":
```
Please remove this block, that's not necessary since quite some time.

Where did you copy this code from?

wagner-certat commented on intelmq/bots/experts/mar/expert_hash.py:
```diff
@@ -0,0 +1,85 @@
+dxl_config_file: string
+
+"""
+from __future__ import unicode_literals
```
We use python3 only anyway

wagner-certat commented on intelmq/bots/collectors/opendxl/collector.py:
```diff
@@ -0,0 +1,106 @@
+dxl_topic: string
+"""
+
+import logging
```
That's not used

wagner-certat commented on intelmq/bots/collectors/opendxl/collector.py:
```diff
@@ -0,0 +1,106 @@
+import sys
+import time
+import json
+import threading
```
not used

wagner-certat commented on intelmq/bots/collectors/opendxl/collector.py:
```diff
@@ -0,0 +1,106 @@
+"""
+
+import logging
+import os
```
not used

wagner-certat commented on intelmq/bots/collectors/opendxl/collector.py:
```diff
@@ -0,0 +1,106 @@
+import time
+import json
+import threading
+import importlib
```
not used

wagner-certat commented on intelmq/bots/collectors/opendxl/collector.py:
```diff
@@ -0,0 +1,106 @@
+from dxlclient.callbacks import EventCallback
+from dxlclient.client import DxlClient
+from dxlclient.client_config import DxlClientConfig
+from dxlclient.message import Event, Request
```
not used

wagner-certat commented on intelmq/bots/experts/mar/expert_hash.py:
```diff
@@ -0,0 +1,85 @@
+
+"""
+from __future__ import unicode_literals
+import sys
```
not used

wagner-certat commented on intelmq/bots/experts/mar/expert_hash.py:
```diff
@@ -0,0 +1,85 @@
+"""
+from __future__ import unicode_literals
+import sys
+import json
```
not used

wagner-certat commented on intelmq/bots/experts/mar/expert_hash.py:
```diff
@@ -0,0 +1,85 @@
+
+from dxlclient.client_config import DxlClientConfig
+from dxlclient.client import DxlClient
+from dxlmarclient import MarClient, ResultConstants, ProjectionConstants, \
```
some of them not used

wagner-certat commented on intelmq/bots/experts/mar/expert_hash.py:
```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+"""
+
+MARHashExpertBot queries environment for occurences of file hashes via McAfee Active Response.
```
Please rename these files to bots/experts/mcafee/expert_mar_hash or similar

wagner-certat commented on intelmq/bots/outputs/esm/output_ip.py:
```diff
@@ -0,0 +1,54 @@
+
+from intelmq.lib.bot import Bot
+
+import json
```
not used

wagner-certat commented on intelmq/bots/outputs/esm/output_ip.py:
```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
```
Please also move this one to `bots/outputs/mcafee/output_esm` or so

wagner-certat commented on intelmq/bots/outputs/esm/output_ip.py:
```diff
@@ -0,0 +1,54 @@
+    def init(self):
+        if ESM is None:
+            self.logger.error('Could not import mfe_saw. Please install it.')
+            self.stop()
```
please use a `raise ValueError('Could not...')` for these two lines instead (I just fixed this in other places too)

wagner-certat commented on intelmq/bots/outputs/esm/output_ip.py:
```diff
@@ -0,0 +1,54 @@
+        self.esm = ESM()
+        try:
+            self.esm.login(self.parameters.esm_ip, self.parameters.esm_user, self.parameters.esm_pw)
+        except:
```
Not every Exception is a failed login. Please catch only a specific exception

wagner-certat commented on intelmq/bots/outputs/esm/output_ip.py:
```diff
@@ -0,0 +1,54 @@
+                    self.logger.info('ESM Watchlist updated')
+                    self.acknowledge_message()
+        except Exception as Err:
+            self.logger.error('Error when updating watchlist.' , Err)
```
use `self.logger.exception('...')` here, it appends the traceback automatically

wagner-certat commented on intelmq/bots/outputs/esm/output_ip.py:
```diff
@@ -0,0 +1,54 @@
+            for WL in retVal:
+                if (WL['name'] == self.parameters.esm_watchlist):
+                    watchlist = {'watchlist': {'value': WL['id']}, 'values': '["' + event.get(self.parameters.field) + '"]'}
+                    print(watchlist)
```
no print please

wagner-certat commented on intelmq/bots/parsers/atd/parser_file.py:
```diff
@@ -0,0 +1,70 @@
+                  severity ranges from 1 to 5
+"""
+from __future__ import unicode_literals
+import sys
```
not used

wagner-certat commented:
Hi,

thanks for the work, 8 new bots for a first time contributor is novelty!

I looked over the code and found some possible improvements. Many comments do apply to similar files too.

I will try to run the code (at least those which are possible for me) this week to test it. Maybe I'll do some cleanups myself.

wagner-certat commented:
And what's missing too:
 * docs how to get access to the feed
 * tests for the parsers (I guess for the other bots tests are not possible)

tux78 commented:
changed code according to coding style, and implemented your recommendations. Documentation of bots/feeds will be updated later on as well. Please let me know if you need any additional assistance for testing

tux78 commented:
I have completed the documentation. However, there still seems to be some issues with it. Maybe you can assist with creating proper documentation.
Thanks,
Seb

wagner-certat commented:
> I have completed the documentation. However, there still seems to be some issues with it. Maybe you can assist with creating proper documentation.

Thanks, will do everything left for the docs.

What I did not yet understand: There is one collector for the sandbox, it get's reports from the "/mcafee/event/atd/file/report" topic. Then there are 3 parsers for files, IPs and URLs. Should the collector feed into all of the parsers or just into one?

tux78 commented:
Thanks for helping with the docs.

Regarding ATD parsing: initial idea was to have one parser per IOC type; but your idea is even better. I am already working on merging ATD Parsers into a single one. I will do the same with the MAR experts, in turn making them more flexible with parameters (e.g.: which IOC you want to process)

tux78 commented:
I picked up the idea of single parser; now reports are generated, one per IOC. I have applied something similar to the MAR expert bot, which is now more flexible.

wagner-certat commented on intelmq/bots/outputs/mcafee/output_esm_ip.py:
```diff
@@ -0,0 +1,59 @@
+        self.esm = ESM()
+        try:
+            self.esm.login(self.parameters.esm_ip, self.parameters.esm_user, self.parameters.esm_pw)
+        except:
```
Which Exception do you expect here?

wagner-certat commented on intelmq/bots/outputs/mcafee/output_esm_ip.py:
```diff
@@ -0,0 +1,59 @@
+                                   watchlist_filter)
+            for WL in retVal:
+                if (WL['name'] == self.parameters.esm_watchlist):
+                    watchlist = {'watchlist': {'value': WL['id']}, 'values': '["' + event.get(self.parameters.field) + '"]'}
```
I looks like you never tested this bot. `event` is not defined. The whole block is part of the wrong function.

wagner-certat commented:
Merged with lots of modifications and adaptions.
First, I squashed many of your small commits as you did not do that yourself.
Second, I repaired the tests as they were failing and some bots. For the latter the modifications were small and as I can't test them myself. The parser was also very broken.
Third, I fixed the codestyle in many places

My commits are in chronological order 666843b70fc19cfc890e5f4e5f10384ab2e45768 d5d1046d8fa900a62fc39c61e8d7ce34259c797c (then you fixed the output) 73ec74c8a80ff792837fd11da7e7f76ebf46f89a 

Please test those bots with real data now, if they behave correctly and do not fail. As I found no way to test those products needed, I can't do that.
